packages/core/saved-objects/core-saved-objects-api-server-internal/src/lib/repository.ts

```diff
@@ -762,6 +773,285 @@ export class SavedObjectsRepository implements ISavedObjectsRepository {
     );
   }
 
+  /**
+   * Performs initial checks on object type validity and flags multi-namespace objects for preflight checks by adding an `esRequestIndex`
+   * @param objects SavedObjectsBulkDeleteObject[]
+   * @returns array BulkDeleteExpectedBulkGetResult[]
+   * @internal
+   */
+  private presortObjectsByNamespaceType(objects: SavedObjectsBulkDeleteObject[]) {
+    let bulkGetRequestIndexCounter = 0;
+    return objects.map<BulkDeleteExpectedBulkGetResult>((object) => {
+      const { type, id } = object;
+      if (!this._allowedTypes.includes(type)) {
+        return {
+          tag: 'Left',
+          value: {
+            id,
+            type,
+            error: errorContent(SavedObjectsErrorHelpers.createUnsupportedTypeError(type)),
+          },
+        };
+      }
+      const requiresNamespacesCheck = this._registry.isMultiNamespace(type);
+      return {
+        tag: 'Right',
+        value: {
+          type,
+          id,
+          ...(requiresNamespacesCheck && { esRequestIndex: bulkGetRequestIndexCounter++ }),
+        },
+      };
+    });
+  }
+
+  /**
+   * Fetch multi-namespace saved objects
```

Comment:
Extracted as a pre-flight check. For now, these helpers are in the main class and can be moved at a later stage to another module.